### PR TITLE
Add additional fields to the UI

### PR DESCRIPTION
**Pull request by burtonr**
Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Description

Adding additional inputs for Environment variables, Secrets,
Labels, and annotations so that more advanced functions can
be deployed through the UI.

This also allows more advanced functions to be provided via
the store. Allowing modification to secret values, and
environment variables which will allow functions like a
Slack Bot function to be deployed through the store adn
configured with an environment variable

Example with the field areas collapsed (default view):

Screenshot from 2019-05-11 16-09-38

Example with the field areas expanded (by clicking on the section title ie: "Environment Variables"):

Screenshot from 2019-05-11 23-21-58

Finally, with all of the fields filled in, showing that multiple values can be added for each group:

Screenshot from 2019-05-11 23-01-24

Now, selecting a function from the store, and switching over to the "Manual" tab will show the additional values that are available (and editable) to the store function as well

Screenshot from 2019-05-11 23-10-04

From the example with all the fields filled out, here is the output of `docker inspect`:
You can see labels, annotations, and Env all available to the created container

```
[
    {
        "ID": "t0mzpevnirqzip2cs1a7v6mdu",
        "Version": {
            "Index": 291
        },
        "CreatedAt": "2019-05-12T03:59:11.0437557Z",
        "UpdatedAt": "2019-05-12T03:59:11.051045016Z",
        "Spec": {
            "Name": "nodeinfo",
            "Labels": {
                "Label1": "LabelValue1",
                "Label2": "LabelValue2",
                "com.openfaas.annotations.Annotate1": "AnnotateValue1",
                "com.openfaas.annotations.Annotate2": "AnnotateValue2",
                "com.openfaas.function": "nodeinfo",
                "function": "true"
            },
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "functions/nodeinfo:latest",
                    "Labels": {
                        "Label1": "LabelValue1",
                        "Label2": "LabelValue2",
                        "com.openfaas.annotations.Annotate1": "AnnotateValue1",
                        "com.openfaas.annotations.Annotate2": "AnnotateValue2",
                        "com.openfaas.function": "nodeinfo",
                        "function": "true"
                    },
                    "Env": [
                        "fprocess=node main.js",
                        "ENV_VAR_2=Variable2",
                        "ENV_VAR_1=Variable1"
                    ],
                    "StopGracePeriod": 10000000000,
                    "DNSConfig": {},
                    "Isolation": "default"
                },
                "Resources": {},
                "RestartPolicy": {
                    "Condition": "any",
                    "Delay": 5000000000,
                    "MaxAttempts": 5
                },
                "Placement": {
                    "Constraints": [
                        "node.platform.os == linux"
                    ]
                },
                "Networks": [
                    {
                        "Target": "kunkx0p11jsyyy7n2luxnkcem"
                    }
                ],
                "ForceUpdate": 0,
                "Runtime": "container"
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            },
            "RollbackConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            }
        },
        "Endpoint": {
            "Spec": {},
            "VirtualIPs": [
                {
                    "NetworkID": "kunkx0p11jsyyy7n2luxnkcem",
                    "Addr": "10.0.1.18/24"
                }
            ]
        }
    }
]

```

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
- [X] My issue has received approval from the maintainers or lead with the `design/approved` label

Discussed in Slack as a "quick-win" with Alex.

## How Has This Been Tested?

Running the OpenFaaS stack locally, with the locally built `openfaas/gateway:latest-dev` image.

Deployed several functions to a local Swarm stack with various additional inputs entered. 
Once deployed, ran `docker service <function name> inspect` to confirm the fields were added properly.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: gateway/assets/script/bootstrap.js [hunk 1]
```diff
@@ -306,6 +329,186 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$interval', '$f
...
+            }
+
+            $scope.removeAnnotation = function($event, annotation) {
+                var index = $scope.annotationInputs.indexOf(annotation);
+                if ($event.which == 1) {
+                    $scope.annotationInputs.splice(index, 1);
+                }
+            }
+
+            $scope.envVarInputsToEnvVars = function() {
```
Comment (burtonr) [inline gateway/assets/script/bootstrap.js]: I created these conversion functions to allow for displaying the key and value of the objects, then translating them back to objects when they're sent to the server. If anyone knows a way to bind input fields to object keys to remove the need for these additional conversions, I'm all ears!

Comment (alexellis) [reply gateway/assets/script/bootstrap.js]: Maybe look into Angular bindings online and see if you can find anything.

**Comment by alexellis** [comment]
What happens if you resize the screen so that the dialog only just fits all the information then keep adding more fields so that the dialog grows beyond the available viewport?

**Comment by burtonr** [comment]
Good question! I forgot to add that screen shot..

The dialog area expands to fill the available vertical space of the window, then beyond that, the dialog becomes scrollable to display all the fields. 

Also, each of the areas can be expanded/contracted to show/hide the information while the input data is retained. That way, if there are a lot of environment variables, you can fill them all out, then hide the area. Additionally, if the fields are populated by the `onFuncSelected()` (aka the function store), the area with values is automatically expanded to show the information that was automatically added.

Screenshot from 2019-05-14 07-46-07

**Comment by LucasRoesler** [comment]
I started testing in in my local docker swarm. Everything seems to be working.  I tried adding invalid values and got the errors I was expecting. It also worked as expected when I submitted with collapsed form sections.   One thing that I did notice is that the form content is preserved after closing the dialog, I would expect the form to be reset when I close it.

**Comment by burtonr** [comment]
Thanks for testing this out @LucasRoesler! 
The form maintaining the state after close was existing functionality. I'm not sure if that was intentional, or not, so I left it as-is.

The form does get reset if you select a different function from the store (again existing functionality)

I'm happy to change that if needed though

**Comment by LucasRoesler** [comment]
It is always best to keep PRs focused, so i would keep it as is for now and open a separate issue (assuming we want to change the behavior) 

Other than that LGTM

**Comment by alexellis** [comment]
I think keeping the state was less confusing before, but now that we can add specific overrides I think it should be fixed.

**Comment by alexellis** [comment]
I think keeping the state was less confusing before, but now that we can add specific overrides I think it should be fixed.

**Comment by burtonr** [comment]
Sounds good! I'll get that updated and pushed tonight. Look for it in the morning :smile:

**Comment by alexellis** [comment]
Thank you

**Comment by alexellis** [comment]
Are we also population envs and labels from the store where they are set?

**Comment by burtonr** [comment]
Yes, when a function is selected from the store, the new fields are populated from the store's definition and passed on to the deployed function, with the ability to adjust them as well.

Although, you would need to know to go to the "Manual" tab after selecting a store function in order to adjust those values. Not sure how we could make that more obvious though.

**Comment by LucasRoesler** [comment]
Could we add a "customize" button to the tile after the function is selected? The button would move the view to the manual tab

**Comment by burtonr** [comment]
The "new function" inputs are now cleared each time the dialog is opened.

I also added a customize button that appears when a function from the store is selected. It shows just to the left of the "deploy" button, so hopefully it will be easy to spot. The color is `md-warn` so it won't get confused with the "deploy" button and can stand out a bit. 

Some screenshots:

Nothing selected:
Screenshot from 2019-05-30 19-04-00

Store function selected:
Screenshot from 2019-05-30 19-03-42

When the "customize" button is clicked, the dialog switches to the "Manual" tab with the values from the store function pre-populated.

**Comment by alexellis** [comment]
Customize sounds new. Perhaps we just change the title of Manual to Custom?

**Comment by burtonr** [comment]
@alexellis good point! Done.

Screenshot from 2019-05-31 07-30-30

**Comment by alexellis** [comment]
Please could you push an updated Docker image that I could test out?

**Comment by burtonr** [comment]
`docker pull burtonr/of-gateway:PR1189`

**Comment by alexellis** [comment]
Screenshot 2019-05-31 at 14 07 45

I'm finding the text "+1" confusing. There are other intuitive symbols we can use for this - i.e. buttons or glyph icons. I'd like to see that changed.

There's also a scrolling problem in Chrome. As more fields are added it's not possible to see a scrollbar or to know that I need to scroll. I can do this with a trackpad, but as a user, if I don't have a mouse with a scrolling-wheel I'm going to be stuck.

**Comment by alexellis** [comment]
I still saw the "CUSTOMIZE" button, can you remove it and run your action when the user clicks "CUSTOM"?

**Comment by alexellis** [comment]
After deploying from the store and clicking "CUSTOM", then launching the dialog again I landed on the "CUSTOM" page, but I'd expect to be brought back to the "FROM STORE" panel when launching after a successful deployment.

**Comment by alexellis** [comment]
Screenshot 2019-05-31 at 14 11 50

I also noticed our timeout values are not in Golang durations right now. @martindekov please could you do a store PR to both JSON files and convert the implicit integer (in seconds) over to a Golang duration format? i.e. `5m` or `300s`.

**Comment by alexellis** [comment]
Screenshot 2019-05-31 at 14 13 50

Just an observation - there doesn't appear to be an option to remove the last / first item in each group?

**Comment by martindekov** [comment]
Yes 👍

**Comment by rgee0** [comment]
ref: https://github.com/openfaas/store/pull/84

**Comment by burtonr** [comment]
> I'm finding the text "+1" confusing. There are other intuitive symbols we can use for this - i.e. buttons or glyph icons. I'd like to see that changed.

The `+1` is a glyph icon from the [material icons](https://material.io/tools/icons/?search=plus&icon=plus_one&style=baseline) (as is the `X`). I looked for something to display there, but thought that a "+1" was most intuitive to adding another field. I'm open to suggestions for something else however.

> There's also a scrolling problem in Chrome. As more fields are added it's not possible to see a scrollbar or to know that I need to scroll. I can do this with a trackpad, but as a user, if I don't have a mouse with a scrolling-wheel I'm going to be stuck.

What version of Chrome are you using? I have testing this in Firefox and Chromium and have not seen the scrollbar missing. You can see it appears in my [previous comment](https://github.com/openfaas/faas/pull/1189#issuecomment-492223665). That may be a problem with the angularjs/material as nothing done here would have affected that. I'm guessing that the function store would have the same issue on your setup since it is the same functionally. 

> I still saw the "CUSTOMIZE" button, can you remove it and run your action when the user clicks "CUSTOM"?

Not sure what you mean here. The "Customize" button appears only when a function is selected from the store. The intention was to draw attention to the fact that there are options to change some of the properties. Are you suggesting that the tab be changed automatically on function selection? If so, I don't think that's a good idea. That would mean if you selected the wrong function, or click several options before being ready to deploy, you would have to change tabs, and re-select what you wanted. Also, not every function needs to be customized.

> After deploying from the store and clicking "CUSTOM", then launching the dialog again I landed on the "CUSTOM" page, but I'd expect to be brought back to the "FROM STORE" panel when launching after a successful deployment.

Should the dialog _always_ open to the store? At the moment, there is logic in there to store the most recent open tab in a variable so that it opens to the same tab that was closed. I guess that was to make it easier for users who only use the "Custom/Manual" tab. Maybe it's best to always open to the store? At least that would make a consistent experience.

> Just an observation - there doesn't appear to be an option to remove the last / first item in each group?

I didn't want to display 2 icons on the first field (add and remove) as it would clutter the view, and also cause some "bouncing" when the `+1` icon disappears/reappears as fields are added/removed. Do we need an icon to clear the text of the input?

**Comment by martindekov** [comment]
golang duration: https://github.com/openfaas/store/pull/85

**Comment by alexellis** [comment]
Here are some alternatives that may be more intuitive:

https://material.io/tools/icons/?search=add&icon=add_box&style=baseline

https://material.io/tools/icons/?search=add&icon=add_circle&style=baseline

https://material.io/tools/icons/?search=add&icon=library_add&style=baseline

**Comment by LucasRoesler** [comment]
I think the circle or the square add icons both look good, in case you need a second opinion :-)

**Comment by alexellis** [comment]
I'd like to hear from Alin, if he has time or Allen.

**Comment by burtonr** [comment]
Before I commit any change, here's what some of the options look like as they will in this UI

**Solid Circles (both)**
Screenshot from 2019-06-07 22-52-59

**Outline Circles (both)**
Screenshot from 2019-06-07 22-55-03

**Circle Add, X Remove**
Screenshot from 2019-06-07 22-53-36


Personally, I'll vote for the outline icons, but they all look pretty nice. What do you think looks best?

**Comment by alexellis** [comment]
Outline circles looks the least alarming on the page. The filled circles remind me of validation hints.

**Comment by burtonr** [comment]
@alexellis I've pushed the outline circles up now after seeing Allen Djal agree with your previous comment.  Let me know if there's anything else remaining on this.

**Comment by alexellis** [comment]
Thank you for leading thus Burton, it's a big step forward. Hope you'll enjoy seeing this being used by all OpenFaaS users going forward!

I'll cut a release then let's get some more people trying to break it :smile:

Alex

**Comment by burtonr** [comment]
Sounds great! Thank you!

**Comment by alexellis** [comment]
https://github.com/openfaas/faas/releases/tag/0.14.0

**Comment by alexellis** [comment]
@burtonr I made a comment about needing to scroll when in overflow. 
Not all people have a trackpad or a scrolling wheel on their mouse

Screenshot 2019-06-16 at 15 16 37

What suggestions do you have for managing this?

**Comment by alexellis** [comment]
I can see why the Customise button may have been added, but I'm not sure it makes sense where it's been placed.

Please can you remove it?

https://github.com/openfaas/faas/pull/1189#issuecomment-497682330

Another approach for a future issue / proposal would to be to add an icon on each Function row to show that it can be customised or edited. We do this already for the repo link button.

Screenshot 2019-06-16 at 15 19 42

**Comment by burtonr** [comment]
Removed the "customize" button on PR #1228 

as for the scrolling issue, I'm unable to reproduce that on anything I have available to me. Even if we set it to auto-focus the new field (which would automatically scroll the window), the fields at the top would then no longer be available. 

There is the option to tab through the fields, which would result in the window scrolling. 

Another option, although more complex, is to create a "wizard style" deployment process. Where each section would get it's own tab with a "next" and "back" button. This would result in less information on the screen at a time. The downside to that is the additional effort when nothing needs changed or added. They would still need to click "next" on up to 4 additional pages...

Do you know of any other site/page where this occurs?

**Comment by alexellis** [comment]
See my screenshot above - taken on Safari/MacOS. I would think it's the same on Chrome/Linux?

**Pull request merged**